CollapseLauncher/Classes/FileMigrationProcess.cs

```diff
@@ -0,0 +1,554 @@
+        {
+            int bufferSize = 1 << 20; // 1MB Buffer
+            if (inputFile.Length < bufferSize)
+                bufferSize = (int)inputFile.Length;
+
+            byte[] buffer = new byte[bufferSize];
+
+            try
+            {
+                await using (FileStream inputStream = inputFile.OpenRead())
+                await using (FileStream outputStream = outputFile.Exists && outputFile.Length <= inputFile.Length ? outputFile.Open(FileMode.Open) : outputFile.Create())
+                {
+                    // Just in-case if the previous move is incomplete, then update and seek to the last position.
+                    if (outputFile.Length <= inputStream.Length && outputFile.Length >= bufferSize)
+                    {
+                        // Do check by comparing the first and last 128K data of the file
+                        Memory<byte> firstCompareInputBytes = new byte[bufferSize];
+                        Memory<byte> firstCompareOutputBytes = new byte[bufferSize];
+                        Memory<byte> lastCompareInputBytes = new byte[bufferSize];
+                        Memory<byte> lastCompareOutputBytes = new byte[bufferSize];
+
+                        // Seek to the first data
+                        inputStream.Position = 0;
+                        await inputStream.ReadExactlyAsync(firstCompareInputBytes);
+                        outputStream.Position = 0;
+                        await outputStream.ReadExactlyAsync(firstCompareOutputBytes);
+
+                        // Seek to the last data
+                        long lastPos = outputStream.Length - bufferSize;
+                        inputStream.Position = lastPos;
+                        await inputStream.ReadExactlyAsync(lastCompareInputBytes);
+                        outputStream.Position = lastPos;
+                        await outputStream.ReadExactlyAsync(lastCompareOutputBytes);
+
+                        bool isMatch = firstCompareInputBytes.Span.SequenceEqual(firstCompareOutputBytes.Span)
+                            && lastCompareInputBytes.Span.SequenceEqual(lastCompareOutputBytes.Span);
+
+                        // If the buffers don't match, then start the copy from the beginning
+                        if (!isMatch)
+                        {
+                            inputStream.Position = 0;
+                            outputStream.Position = 0;
+                        }
+                        else
+                        {
+                            UpdateSizeProcessed(uiRef, outputStream.Length);
+                        }
+                    }
+
+                    await MoveWriteFileInner(uiRef, inputStream, outputStream, buffer, token);
+                }
+
+                inputFile.IsReadOnly = false;
+                inputFile.Delete();
+            }
+            catch { throw; }
```

Comment:
empty catch-throw-finally is generally not desirable as it will cause us and user confusion when debugging errors. might want to add at least a Logger line or better yet, an ErrorSender.

Reply:
This should be redundant since the exception handling will happen at ``InstallManagerBase`` after all.
But I expected to write it that way since I need to handle finalization on the use of ``ArrayPool<byte>`` and to handle the exception by rethrowing it, which I haven't implement yet in this commit and will be included in the next commit.